### DISCUSSION TITLE
Algorithm of the original version

### DISCUSSION THREAD

**Discussion by joaomcteixeira** [THGLab/MCSCE#11] [Implementations]
Use this thread to detail the algorithm of the original version. Be as detailed as possible.

1. entry point `mcsce.py`
    * Accepts arguments (`file`, `-n`, `-t`, `-o`)
    * Assert PDB has only one model and one chain.
    * Counts the number of residues in the PDB file.
        * If PDB has no residues, fails
    * Creates the output folder based on `-o`
    * Copies the template scripts to the output folder (:thought_balloon: this may be avoided in the new implementation)
    * In the new script files copied to the output directory:
        * Replace the `resnum` part in `prepareSubmission` script
        * Set number of trials in `setupentropy` script
        * Specify script folder position in `setupentropy` script
    * `bash` run the new `prepareSubmission` script.
        * args:
            * path to PDB file
            * path to output dir
1. `prepareSubmission` script
    * Takes 3 arguments (round, pdb, nfile)
       * round: "pdb"
       * state: "lig"
       * pdb: input pdb file path
       * nfile: number of generated output files
       * nres: number of residues of the protein
    * copy the original pdb file into 0.pdb
    * do nfile times: 
       * make working folder for the nth structure
       * copy `setupentropy.csh`, all `lig2.*` files into working folder and copy the original pdb file as `${round}.pdb` (pdb.pdb) 
       ```
       · find . -name lig2*
       ./template_lig/lig2.mol2
       ./template_lig/lig2.lib
       ./template_lig/lig2.pdb
       ./template_lig/lig2.frcmod
       ```
       * change all `HETATM` to `ATOM  `  in pdb files
       * only keep lines that correspond to non-H atoms (the original implementation is actually WRONG!!! It also removes some non-hydrogen atoms with letter "H" in atom name) (:thought_balloon: well spotted, this is very easy to solve)
           * [Uses `awk` in `bash`](https://pastebin.com/eQbbAatX), :thought_balloon: but can be done in Python using the PDB parser of choice.
       * create `ptraj.in` file
         ```
         trajin pdb.inpcrd     # pdb comes from ${round}
         trajout traj.1 pdb
      * create `leap.in` file
         ```
         # all "pdb" here are replaced from ${round}
         source /share/apps/software/amber16/dat/leap/cmd/oldff/leaprc.ff99SB     
         source /share/apps/software/amber16/dat/leap/cmd/leaprc.gaff
         loadamberparams frcmod.tip4pew
         loadamberparams lig2.frcmod
         loadoff lig2.lib
         HOH = T4E
         WAT = T4E
         pdb = loadpdb pdb.pdb
         saveamberparm pdb pdb.prmtop pdb.inpcrd
         quit
      * execute AMBER preprocessing
        * :thought_balloon: Amber can ben installed now via Anaconda, at least the simplest version. The new Amber20 also has a new installation method [link](http://ambermd.org/GetAmber.php). Can we use this to avoid reimplementations?
        * `$AMBERHOME/tleap -f leap.in`
        * `$AMBERHOME/cpptraj pdb.prmtop ptraj.in`
      * execute `setupentropy.csh` script with following parameters:
        * pdb (${round})
        * lig (${state})
        * number of residues (${nres})
        * current calculation structure index (${i})
      * execute `submit.bash` for submitting bash job for cluster calculation 

 1. `setupentropy.csh` script
    * Takes 4 arguments:
       * prefix: filename prefix
       * state: either "apo" or "lig"
       * ctres: C-terminus residue index (Assuming N-terminus index is 1)
       * output_name: output filename
    * create `input` file
      ```
      ##
      PDBINPUT             traj_3.pdb
      PDBOUTPUT          traj_4.pdb
      XYZOUTPUT            xxx
      AMBER2TINKERTABLE    MCSCE_PATH/script_lig/sample.entropy/amber2tinker.table
      NTERIDX              1
      CTERIDX              ${ctres}
      ##
    * Keep only atom lines (start with ATOM/HETATM) that do not contain HOH/WAT/Na+/Cl-, remove occupancy, b-factor and other non-necessary PDB fields, and rename all HIP into HIS. The new pdb is saved to `traj_1.pdb`.
    * Remove hydrogen using `script_lig/removeH.pl`
    * Process PDB using `scruot_lig/processPDBwithH.pl` and generates `traj_3.pdb` :question: Not sure what are done exactly in this perl file
    * Depending on `${state}` equals to lig, get all lines from `traj_1.pdb` with "LIG" in it, replace all occurance of "LIG" into "LG1" and append these lines to the end of `traj_3.pdb`
    * Run `script_lig/pdb2zpdb.pl` on `input` to generate `traj_4.pdb`, and soft link the file into `ent.pdb`  :question: Not sure what are done exactly in this perl file
    * Create `tinker.key` file
    * Finally create `submit.bash` for submitting jobs to the cluster
    * The aim of `setupentropy.csh` is to make a script that ultimately runs `MCSCE_PATH/source2.0303_v13.9/modbuilder3.x`.
1. `tinker.key`
This is the Tinker input key file which defines the parameter to use (amber99sb), dielectric (=4.0), and a bunch of SCE related settings:
number of trials (500), temperature (300), solvation (GB-HPMF), definition of flexible atoms in different residues, and rotamer libraries to use 
1. `submit.bash`
This script uses mpirun to execute `modbuilder3.x`

**Comment by joaomcteixeira** [comment]
:question:

> change all HETATM to ATOM in pdb files

Is this done to achieve compatibility with Amber?

**Reply by JerryJohnsonLee** [reply]
I believe so. Maybe Amber does not recognize HETATM? I am not sure because I have no experience running Amber
